pybamm/solvers/jax_bdf_solver.py

```diff
@@ -0,0 +1,719 @@
+    y_out, stepper = _bdf_odeint(fun, jac, rtol, atol, y0_device, t_eval_device, inputs)
+    return y_out, stepper
+
+
+MAX_ORDER = 5
+NEWTON_MAXITER = 4
+MIN_FACTOR = 0.2
+MAX_FACTOR = 10
+
+
+def flax_cond(pred, true_operand, true_fun,
+              false_operand, false_fun):  # pragma: no cover
+    """
+    for debugging purposes, use this instead of jax.lax.cond
+    """
+    if pred:
+        return true_fun(true_operand)
+    else:
+        return false_fun(false_operand)
+
+
+def flax_while_loop(cond_fun, body_fun, init_val):  # pragma: no cover
+    """
+    for debugging purposes, use this instead of jax.lax.while_loop
+    """
+    val = init_val
+    while cond_fun(val):
+        val = body_fun(val)
+    return val
+
+
+def flax_fori_loop(start, stop, body_fun, init_val):  # pragma: no cover
+    """
+    for debugging purposes, use this instead of jax.lax.fori_loop
+    """
+    val = init_val
+    for i in range(start, stop):
+        val = body_fun(i, val)
+    return val
+
+
+def _compute_R(order, factor):
+    """
+    computes the R matrix with entries
+    given by the first equation on page 8 of [1]
+
+    This is used to update the differences matrix when step size h is varied according
+    to factor = h_{n+1} / h_n
+
+    Note that the U matrix also defined in the same section can be also be
+    found using factor = 1, which corresponds to R with a constant step size
+    """
+    I = np.arange(1, MAX_ORDER + 1).reshape(-1, 1)
+    J = np.arange(1, MAX_ORDER + 1)
+    M = np.empty((MAX_ORDER + 1, MAX_ORDER + 1))
+    M = jax.ops.index_update(M, jax.ops.index[1:, 1:],
+                             (I - 1 - factor * J) / I)
+    M = jax.ops.index_update(M, jax.ops.index[0], 1)
+    R = np.cumprod(M, axis=0)
+
+    return R
+
+
+def _bdf_init(fun, jac, t0, y0, h0, rtol, atol):
+    """
+    Backward Difference formula (BDF) implicit multistep integrator. The basic algorithm
+    is derived in [2]. This particular implementation follows that implemented in the
+    Matlab routine ode15s, described in [1], which features the NDF formulas for
+    improved stability, with associated differences in the error constants, and
+    calculates the jacobian at J(t_n, y_n), rather than at the more standard J(t_{n+1},
+    y^0_{n+1}). This implementation was based on that implemented in the scipy library
+    [3], which also mainly follows [1] but uses the more standard jacobian update.
+
+    Parameters
+    ----------
+
+    fun: callable
+        function with signature (t, y), where t is a scalar time and y is a ndarray with
+        shape (n,), returns the rhs of the system of ODE equations as an nd array with
+        shape (n,)
+    jac: callable
+        function with signature (t, y), where t is a scalar time and y is a ndarray with
+        shape (n,), returns the jacobian matrix of fun as an ndarray with shape (n,n)
+    t0: float
+        initial time
+    y0: ndarray
+        initial state vector with shape (n,)
```

Comment:
missing rtol and atol docs

Reply:
forgot to update that docstring, done now